src/login/logind.c

```diff
@@ -444,74 +444,80 @@ static int parse_fdname(const char *fdname, char **session_id, dev_t *dev) {
+                goto close;
+        }
 
-        for (i = 0; i < n; i++) {
-                _cleanup_free_ char *id = NULL;
-                dev_t dev;
-                struct stat st;
-                SessionDevice *sd;
-                Session *s;
-                int r;
+        if (fstat(fd, &st) < 0) {
+                /* The device is allowed to go away at a random point, in which
+                 * case fstat failing is expected. */
+                log_debug_errno(errno, "Failed to stat device fd for session %s: %m", id);
+                goto close;
+        }
 
-                fd = SD_LISTEN_FDS_START + i;
+        if (!S_ISCHR(st.st_mode) || st.st_rdev != dev) {
+                log_debug("Device fd doesn't point to the expected character device node");
+                goto close;
+        }
 
-                r = parse_fdname(fdnames[i], &id, &dev);
-                if (r < 0) {
-                        log_debug_errno(r, "Failed to parse fd name %s: %m", fdnames[i]);
-                        close_nointr(fd);
-                        continue;
-                }
+        sd = hashmap_get(s->devices, &dev);
+        if (!sd) {
+                /* Weird, we got an fd for a session device which wasn't
+                 * recorded in the session state file... */
+                log_warning("Got fd for missing session device [%u:%u] in session %s",
+                            major(dev), minor(dev), s->id);
+                goto close;
+        }
 
-                s = hashmap_get(m->sessions, id);
-                if (!s) {
-                        /* If the session doesn't exist anymore, the associated session
-                         * device attached to this fd doesn't either. Let's simply close
-                         * this fd. */
-                        log_debug("Failed to attach fd for unknown session: %s", id);
-                        close_nointr(fd);
-                        continue;
-                }
+        log_debug("Attaching fd to session device [%u:%u] for session %s",
+                  major(dev), minor(dev), s->id);
 
-                if (fstat(fd, &st) < 0) {
-                        /* The device is allowed to go away at a random point, in which
-                         * case fstat failing is expected. */
-                        log_debug_errno(errno, "Failed to stat device fd for session %s: %m", id);
-                        close_nointr(fd);
-                        continue;
-                }
+        session_device_attach_fd(sd, fd, s->was_active);
+        return;
 
-                if (!S_ISCHR(st.st_mode) || st.st_rdev != dev) {
-                        log_debug("Device fd doesn't point to the expected character device node");
-                        close_nointr(fd);
-                        continue;
-                }
+close:
+        r = asprintf(&message, "FDSTOREREMOVE=1\n"
+                               "FDNAME=%s\n",
+                               fdname);
```

Comment:
just use sd_notifyf() please!

Reply:
Wow, we managed to miss that lovely convenience function earlier.  Thanks!